glean-core/android/src/main/java/mozilla/telemetry/glean/Dispatchers.kt

```diff
@@ -45,6 +47,68 @@ internal object Dispatchers {
         }
     }
 
+    class DelayedTaskQueue {
+        // When true, jobs will be queued and not ran until triggered by calling
+        // flushQueuedInitialTasks()
+        private var queueInitialTasks = AtomicBoolean(true)
+
+        // Use a [ConcurrentLinkedQueue] to take advantage of its thread safety and no locking
+        internal val taskQueue: ConcurrentLinkedQueue<() -> Unit> = ConcurrentLinkedQueue()
+
+        /**
+         * Launch a block of work synchronously or queue it if not yet unblocked.
+         *
+         * If [queueInitialTasks] true
+         * then the work will be queued and executed when [flushQueuedInitialTasks] is called.
+         * If [queueInitialTasks] is false the block is executed synchronously.
+         */
+        fun launch(
+            block: () -> Unit,
+        ) {
+            if (queueInitialTasks.get()) {
+                addTaskToQueue(block)
+            } else {
+                block()
+            }
+        }
+
+        /**
+         * Stops queueing tasks and processes any tasks in the queue.
+         *
+         * Processing happens on the current thread synchronously.
+         *
+         * Since [queueInitialTasks] is set to false prior to processing the queue,
+         * newly launched tasks should be executed immediately rather than added to the queue.
+         */
+        internal fun flushQueuedInitialTasks() {
+            val dispatcherObject = this
+
+            queueInitialTasks.set(false)
+
+            // Nothing should be added to this list. However, the flush could get called
+            // while a `addTaskToQueue` is being executed. For this reason, we need synchronized
+            // access to the queue. However, we can't simply wrap the task execution in a sync
+            // block: suspending functions are not allowed inside critical sections.
+            val queueCopy: ConcurrentLinkedQueue<() -> Unit> = ConcurrentLinkedQueue()
+            synchronized(dispatcherObject) {
+                queueCopy.addAll(taskQueue)
+                taskQueue.clear()
+            }
+
+            queueCopy.forEach { task ->
```

Comment:
I assume this iterates through the queue in a FIFO way?

Reply:
Yup, this is the same code as the old dispatcher used.

Reply:
Sorry, I guess I must have blocked that out 😅